diff --git a/meson.build b/meson.build
index d034f33..8bf3c30 100644
--- a/meson.build
+++ b/meson.build
@@ -16,9 +16,7 @@ gee_dependency = dependency ('gee-0.8')
 gsound_dependency = dependency('gsound', version: '>= 1.0.2')
 gtk_dependency = dependency('gtk4', version: '>= 4.4')
 libadwaita_dependency = dependency('libadwaita-1', version: '>= 1.1')
-libgnome_games_support = subproject ('libgnome-games-support',
-	    		             default_options: 'superproject_name=quadrapassel')
-libgnome_games_support_dependency = libgnome_games_support.get_variable ('lggs_dependency')
+libgnome_games_support_dependency = dependency ('libgnome-games-support-3', version: '>= 3.0.0')
 manette_dependency = dependency('manette-0.2', version: '>= 0.2.0')
 pango_dependency = dependency('pango')
 pango_cairo_dependency = dependency('pangocairo')
