diff --git a/js/ui/main.js b/js/ui/main.js
index 76b917f48c..1fc75e130f 100644
--- a/js/ui/main.js
+++ b/js/ui/main.js
@@ -271,6 +271,7 @@ async function _initializeUI() {
             timeLimitsManager.shutdown()
                 .catch(e => console.warn(`Failed to stop time limits manager: ${e.message}`))
                 .finally(() => loop.quit());
+        _notifyGnomeSessionReady();
             return GLib.SOURCE_REMOVE;
         });
         source.attach(loop.get_context());
@@ -379,6 +380,20 @@ async function _initializeUI() {
     });
 }
 
+async function _notifyGnomeSessionReady() {
+    try {
+        let params = GLib.Variant.new('(ss)', ['org.gnome.Shell.desktop', '']);
+        await Gio.DBus.session.call(
+            'org.gnome.SessionManager',
+            '/org/gnome/SessionManager',
+            'org.gnome.SessionManager',
+            'RegisterClient', params, null,
+            Gio.DBusCallFlags.NONE, -1, null);
+    } catch (e) {
+        log(`Error notifying gnome-session that we're ready: ${e.message}`);
+    }
+}
+
 function _handleShowWelcomeScreen() {
     const lastShownVersion = global.settings.get_string(WELCOME_DIALOG_LAST_SHOWN_VERSION);
     if (Util.GNOMEversionCompare(WELCOME_DIALOG_LAST_TOUR_CHANGE, lastShownVersion) > 0) {
