From 38b7dc34b9a78629d02bf6e4e3a1b95c66b49120 Mon Sep 17 00:00:00 2001
From: "W. Michael Petullo" <mike@flyn.org>
Date: Wed, 14 Dec 2022 10:28:16 -0600
Subject: [PATCH] libdmapsharing/Makefile.am: replace Soup-2.4 with Soup-3.0

Signed-off-by: W. Michael Petullo <mike@flyn.org>
---
 libdmapsharing/Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libdmapsharing/Makefile.am b/libdmapsharing/Makefile.am
index cd55ec7..7f4e36c 100644
--- a/libdmapsharing/Makefile.am
+++ b/libdmapsharing/Makefile.am
@@ -181,7 +181,7 @@ Dmap_4_0_gir_NAMESPACE = Dmap
 Dmap_4_0_gir_SCANNERFLAGS = --symbol-prefix=dmap
 Dmap_4_0_gir_LIBS = libdmapsharing-4.0.la
 Dmap_4_0_gir_FILES = $(introspection_sources)
-Dmap_4_0_gir_INCLUDES = GObject-2.0 Soup-2.4 Gst-1.0
+Dmap_4_0_gir_INCLUDES = GObject-2.0 Soup-3.0 Gst-1.0
 Dmap_4_0_gir_CFLAGS = $(INCLUDES) $(libdmapsharing_4_0_la_CFLAGS)
 
 INTROSPECTION_GIRS = Dmap-4.0.gir
-- 
GitLab

