From e3bb220cdff066a9354c96b7db5e4849091ad6c8 Mon Sep 17 00:00:00 2001
From: Mat <mail@mathias.is>
Date: Tue, 4 Feb 2025 05:38:08 +0200
Subject: [PATCH] score-dialog: Cast sort callbacks to CompareDataFunc

---
 src/score-dialog.vala | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/src/score-dialog.vala b/src/score-dialog.vala
index 299ef6b0..dd88482d 100644
--- a/src/score-dialog.vala
+++ b/src/score-dialog.vala
@@ -185,9 +185,9 @@ public class ScoreDialog : Adw.Dialog {
 
             label.label = (position + 1).to_string ();
         });
-        sorter.append (new Gtk.CustomSorter (rank_sorter_cb));
-        sorter.append (new Gtk.CustomSorter (date_sorter_cb));
-        sorter.append (new Gtk.CustomSorter (player_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) rank_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) date_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) player_sorter_cb));
 
         rank_column.sorter = sorter;
         rank_column.factory = factory;
@@ -246,9 +246,9 @@ public class ScoreDialog : Adw.Dialog {
                 inscription.text = entry.player;
             }
         });
-        sorter.append (new Gtk.CustomSorter (player_sorter_cb));
-        sorter.append (new Gtk.CustomSorter (rank_sorter_cb));
-        sorter.append (new Gtk.CustomSorter (date_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) player_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) rank_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) date_sorter_cb));
 
         player_column.sorter = sorter;
         player_column.factory = factory;
@@ -304,9 +304,9 @@ public class ScoreDialog : Adw.Dialog {
 
             label.label = date_label;
         });
-        sorter.append (new Gtk.CustomSorter (date_sorter_cb));
-        sorter.append (new Gtk.CustomSorter (rank_sorter_cb));
-        sorter.append (new Gtk.CustomSorter (player_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) date_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) rank_sorter_cb));
+        sorter.append (new Gtk.CustomSorter ((CompareDataFunc<HistoryEntry>) player_sorter_cb));
 
         date_column.sorter = sorter;
         date_column.factory = factory;
-- 
GitLab

