--- gsettings-desktop-schemas-42.0/schemas/org.gnome.desktop.interface.gschema.xml.in   2022-03-19 22:11:55.000000000 +0000
+++ org.gnome.desktop.interface.gschema.xml.in  2022-06-07 03:42:33.366000000 +0100
@@ -163,7 +163,7 @@
       </description>
     </key>
     <key name="monospace-font-name" type="s">
-      <default>'Source Code Pro 10'</default>
+      <default>'Monospace 11'</default>
       <summary>Monospace font</summary>
       <description>
         Name of a monospaced (fixed-width) font for use in locations like
