diff --git a/CMakeLists.txt b/CMakeLists.txt
index cb754134..2fddd2bb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -184,7 +184,7 @@ if(WIN32)
   set_target_properties( ${ly_lib_shared} PROPERTIES IMPORT_PREFIX "lib" )
 endif()
 
-SET_TARGET_PROPERTIES   ( ${ly_lib_shared} PROPERTIES VERSION "0.0.0" SOVERSION 0 )
+set_target_properties   ( ${ly_lib_shared} PROPERTIES VERSION "0.0.1904" SOVERSION 0 )
 
 # this creates the cpuid tool
 add_executable      ( cpuid ${ly_base_dir}/util/cpuid.c )

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2fddd2bb..fa2590ba 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -196,7 +196,7 @@ target_link_libraries	( yuvconvert ${ly_lib_shared} )
 
 # this creates the yuvconstants tool
 add_executable      ( yuvconstants ${ly_base_dir}/util/yuvconstants.c )
-target_link_libraries  ( yuvconstants ${ly_lib_shared} )
+target_link_libraries  ( yuvconstants ${ly_lib_shared} m )
 
 find_package ( JPEG )
 if (JPEG_FOUND)
