diff --git a/data/gnome-mimeapps.list b/data/gnome-mimeapps.list
index cf6b21b0..1f558f0a 100644
--- a/data/gnome-mimeapps.list
+++ b/data/gnome-mimeapps.list
@@ -49,7 +49,6 @@ application/x-font-otf=org.gnome.font-viewer.desktop
 application/x-font-pcf=org.gnome.font-viewer.desktop
 application/x-font-ttf=org.gnome.font-viewer.desktop
 application/x-font-type1=org.gnome.font-viewer.desktop
-application/x-freedesktop-appstream-component=gnome-software-local-file-metainfo.desktop
 application/x-gnome-app-info=org.gnome.Builder.desktop
 application/x-gzip=org.gnome.Nautilus.desktop
 application/x-gzpdf=org.gnome.Papers.desktop
@@ -69,8 +68,8 @@ application/x-php-source=org.gnome.Builder.desktop
 application/x-pkcs12=gcr-viewer.desktop
 application/x-pkcs7-certificates=gcr-viewer.desktop
 application/x-python=org.gnome.Builder.desktop
-application/x-raw-disk-image=gnome-disk-image-writer.desktop;gnome-disk-image-mounter.desktop
-application/x-raw-disk-image-xz-compressed=gnome-disk-image-writer.desktop;gnome-disk-image-mounter.desktop
+application/x-raw-disk-image=gnome-disk-image-mounter.desktop
+application/x-raw-disk-image-xz-compressed=gnome-disk-image-mounter.desktop
 application/x-ruby=org.gnome.Builder.desktop
 application/x-shellscript=org.gnome.Builder.desktop
 application/x-spkac=gcr-viewer.desktop
@@ -115,37 +114,32 @@ audio/x-speex=org.gnome.Decibels.desktop
 audio/x-vorbis=org.gnome.Decibels.desktop
 audio/x-vorbis+ogg=org.gnome.Decibels.desktop
 audio/x-wavpack=org.gnome.Decibels.desktop
-font/collection=org.gnome.font-viewer.desktop
 font/otf=org.gnome.font-viewer.desktop
 font/ttf=org.gnome.font-viewer.desktop
 font/woff=org.gnome.font-viewer.desktop
-image/apng=org.gnome.Loupe.desktop
 image/avif=org.gnome.Loupe.desktop
 image/bmp=org.gnome.Loupe.desktop
 image/gif=org.gnome.Loupe.desktop
 image/heic=org.gnome.Loupe.desktop
-image/jp2=org.gnome.Loupe.desktop
 image/jpeg=org.gnome.Loupe.desktop
 image/jxl=org.gnome.Loupe.desktop
 image/png=org.gnome.Loupe.desktop
 image/qoi=org.gnome.Loupe.desktop
-image/svg+xml=org.gnome.Loupe.desktop;org.gtk.Shaper.desktop
+image/svg+xml=org.gnome.Loupe.desktop
 image/svg+xml-compressed=org.gnome.Loupe.desktop
 image/tiff=org.gnome.Loupe.desktop;org.gnome.Papers.desktop
+image/vnd-ms.dds=org.gnome.Loupe.desktop
 image/vnd.microsoft.icon=org.gnome.Loupe.desktop
+image/vnd.radiance=org.gnome.Loupe.desktop
 image/webp=org.gnome.Loupe.desktop
 image/x-dds=org.gnome.Loupe.desktop
 image/x-exr=org.gnome.Loupe.desktop
-image/x-gtk-path-animation=org.gtk.Shaper.desktop
 image/x-portable-anymap=org.gnome.Loupe.desktop
 image/x-portable-bitmap=org.gnome.Loupe.desktop
 image/x-portable-graymap=org.gnome.Loupe.desktop
 image/x-portable-pixmap=org.gnome.Loupe.desktop
 image/x-qoi=org.gnome.Loupe.desktop
 image/x-tga=org.gnome.Loupe.desktop
-image/x-win-bitmap=org.gnome.Loupe.desktop
-image/x-xbitmap=org.gnome.Loupe.desktop
-image/x-xpixmap=org.gnome.Loupe.desktop
 inode/directory=org.gnome.Nautilus.desktop
 message/rfc822=org.gnome.Epiphany.desktop
 multipart/related=org.gnome.Epiphany.desktop
@@ -243,8 +237,6 @@ x-scheme-handler/snap=org.gnome.Software.desktop
 x-scheme-handler/tel=org.gnome.Calls.desktop
 
 # Tracking data to catch stale overrides:
-#OVERRIDE application/x-raw-disk-image WAS gnome-disk-image-mounter.desktop
-#OVERRIDE application/x-raw-disk-image-xz-compressed WAS gnome-disk-image-mounter.desktop
 #OVERRIDE image/tiff WAS org.gnome.Loupe.desktop org.gnome.Papers.desktop
 #OVERRIDE inode/directory WAS org.gnome.Nautilus.desktop org.gnome.baobab.desktop
 #OVERRIDE text/html WAS org.gnome.Builder.desktop org.gnome.Epiphany.desktop
diff --git a/data/gnome-session-basic-services.target b/data/gnome-session-basic-services.target
index 84f9e47d..96710c6a 100644
--- a/data/gnome-session-basic-services.target
+++ b/data/gnome-session-basic-services.target
@@ -14,6 +14,8 @@ Requisite=gnome-session.target
 PartOf=gnome-session.target
 Before=gnome-session.target
 
+Requires=org.gnome.Shell.target
+
 # Manages a11y settings, starts Orca when needed
 Wants=org.gnome.SettingsDaemon.A11ySettings.target
 
diff --git a/data/gnome-session-services.target b/data/gnome-session-services.target
deleted file mode 100644
index efe25aac..00000000
--- a/data/gnome-session-services.target
+++ /dev/null
@@ -1,63 +0,0 @@
-[Unit]
-Description=GNOME session services
-OnFailure=gnome-session-shutdown.target
-OnFailureJobMode=replace-irreversibly
-DefaultDependencies=no
-RefuseManualStart=yes
-RefuseManualStop=yes
-
-Requisite=gnome-session-initialized.target
-After=gnome-session-initialized.target
-PartOf=gnome-session-initialized.target
-
-Requisite=gnome-session.target
-PartOf=gnome-session.target
-Before=gnome-session.target
-
-# Basic services required by any GNOME session
-Requires=gnome-session-basic-services.target
-
-# Manages monitor color calibration and night-light
-Wants=org.gnome.SettingsDaemon.Color.target
-
-# Manages location-based timezone
-Wants=org.gnome.SettingsDaemon.Datetime.target
-
-# Notifies when disk is full, notifies when RAM is full,
-# cleans up trash and temp files, donation reminders
-Wants=org.gnome.SettingsDaemon.Housekeeping.target
-
-# Notifies when events happen with printers: low on ink,
-# jammed, job cancelled, etc etc etc
-Wants=org.gnome.SettingsDaemon.PrintNotifications.target
-
-# org.freedesktop.ScreenSaver API
-Wants=org.gnome.SettingsDaemon.ScreensaverProxy.target
-
-# Flushes PulseAudio sample cache whenever the
-# sound theme changes
-Wants=org.gnome.SettingsDaemon.Sound.target
-
-# Manages SIM cards (unlocking, etc)
-Wants=org.gnome.SettingsDaemon.Wwan.target
-
-# Monitors disks for SMART errors, and notifies
-# the user if their disk is about to fail
-Wants=gnome-disk-utility-notify.service
-
-# Filesystem indexer
-Wants=localsearch-3.service
-
-# Graphical wrapper around ssh-agent
-Wants=gcr-ssh-agent.socket
-
-# Checking for automatic updates, etc
-Wants=gnome-software.service
-
-# Translates the user's standard directories
-# (Downloads/, Documents/, Music/, etc) whenever
-# the user's locale changes
-Wants=user-dirs-update-gtk.service
-
-# Notifies the user about their calendar events
-Wants=evolution-alarm-notify.service
diff --git a/data/gnome-session-wayland.target b/data/gnome-session-wayland.target
new file mode 100644
index 00000000..9a0a72a5
--- /dev/null
+++ b/data/gnome-session-wayland.target
@@ -0,0 +1,12 @@
+[Unit]
+Description=GNOME Wayland Session
+OnFailure=gnome-session-shutdown.target
+OnFailureJobMode=replace-irreversibly
+# Avoid default After/Before rules
+DefaultDependencies=no
+
+Before=gnome-session.target
+
+PartOf=graphical-session.target
+RefuseManualStart=yes
+RefuseManualStop=yes
diff --git a/data/gnome-session-wayland@.target b/data/gnome-session-wayland@.target
new file mode 100644
index 00000000..ba2a9185
--- /dev/null
+++ b/data/gnome-session-wayland@.target
@@ -0,0 +1,20 @@
+[Unit]
+Description=GNOME Wayland Session (session: %i)
+DefaultDependencies=no
+# Start happens explicitly
+RefuseManualStart=no
+# Stop happens by starting gnome-session-shutdown.target
+RefuseManualStop=yes
+
+Conflicts=shutdown.target gnome-session-shutdown.target
+PartOf=graphical-session.target
+
+# As this is the main entry point, pull in the other toplevel gnome-session targets
+Requires=gnome-session@.target
+After=gnome-session@.target
+
+Requires=gnome-session-wayland.target
+After=gnome-session-wayland.target
+
+Requires=gnome-session.target
+After=gnome-session.target
diff --git a/data/gnome-session-x11.target b/data/gnome-session-x11.target
new file mode 100644
index 00000000..a91d0877
--- /dev/null
+++ b/data/gnome-session-x11.target
@@ -0,0 +1,15 @@
+[Unit]
+Description=GNOME X11 Session
+OnFailure=gnome-session-shutdown.target
+OnFailureJobMode=replace-irreversibly
+# Avoid default After/Before rules
+DefaultDependencies=no
+
+Before=gnome-session.target
+
+PartOf=graphical-session.target
+RefuseManualStart=yes
+RefuseManualStop=yes
+
+# Pull in all X11-specific services the session might depend on
+Requires=gnome-session-x11-services.target
diff --git a/data/gnome-session-x11@.target b/data/gnome-session-x11@.target
new file mode 100644
index 00000000..b371759a
--- /dev/null
+++ b/data/gnome-session-x11@.target
@@ -0,0 +1,22 @@
+[Unit]
+Description=GNOME X11 Session (session: %i)
+OnFailure=gnome-session-shutdown.target
+OnFailureJobMode=replace-irreversibly
+DefaultDependencies=no
+# Start happens explicitly
+RefuseManualStart=no
+# Stop happens by starting gnome-session-shutdown.target
+#RefuseManualStop=yes
+
+Conflicts=shutdown.target gnome-session-shutdown.target
+PartOf=graphical-session.target
+
+# As this is the main entry point, pull in the other toplevel gnome-session targets
+BindsTo=gnome-session@.target
+After=gnome-session@.target
+
+BindsTo=gnome-session-x11.target
+After=gnome-session-x11.target
+
+BindsTo=gnome-session.target
+After=gnome-session.target
diff --git a/data/gnome-session@.target b/data/gnome-session@.target
index dc312fa5..d5ca3a11 100644
--- a/data/gnome-session@.target
+++ b/data/gnome-session@.target
@@ -1,21 +1,16 @@
 [Unit]
 Description=GNOME Session (session: %i)
-DefaultDependencies=no
-# Start happens explicitly
-RefuseManualStart=no
-# Stop happens by starting gnome-session-shutdown.target
-RefuseManualStop=yes
-
 OnFailure=gnome-session-shutdown.target
 OnFailureJobMode=replace-irreversibly
-
-Conflicts=shutdown.target gnome-session-shutdown.target
-PartOf=graphical-session.target
+DefaultDependencies=no
+RefuseManualStart=yes
+RefuseManualStop=yes
 
 Requires=gnome-session-initialized.target
 After=gnome-session-initialized.target
 
-Requires=gnome-session.target
+Requisite=gnome-session.target
 PartOf=gnome-session.target
+Before=gnome-session.target
 
 Requires=gnome-session-manager@.service
diff --git a/data/gnome-wayland.desktop.in.in b/data/gnome-wayland.desktop.in.in
new file mode 100644
index 00000000..b747802d
--- /dev/null
+++ b/data/gnome-wayland.desktop.in.in
@@ -0,0 +1,9 @@
+[Desktop Entry]
+Name=GNOME on Wayland
+Comment=This session logs you into GNOME
+Exec=@bindir@/gnome-session
+TryExec=@bindir@/gnome-session
+Type=Application
+DesktopNames=GNOME
+X-GDM-SessionRegisters=true
+X-GDM-CanRunHeadless=true
diff --git a/data/gnome-xorg.desktop.in.in b/data/gnome-xorg.desktop.in.in
new file mode 100644
index 00000000..9a76fac8
--- /dev/null
+++ b/data/gnome-xorg.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=GNOME on Xorg
+Comment=This session logs you into GNOME
+Exec=@bindir@/gnome-session
+TryExec=@bindir@/gnome-session
+Type=Application
+DesktopNames=GNOME
+X-GDM-SessionRegisters=true
diff --git a/data/gnome.desktop.in.in b/data/gnome.desktop.in.in
index d6c228ae..e3d4edaf 100644
--- a/data/gnome.desktop.in.in
+++ b/data/gnome.desktop.in.in
@@ -6,4 +6,4 @@ TryExec=@bindir@/gnome-session
 Type=Application
 DesktopNames=GNOME
 X-GDM-SessionRegisters=true
-X-GDM-CanRunHeadless=true
+X-GDM-CanRunHeadless=@canrunheadless@
diff --git a/data/gnome.session.conf b/data/gnome.session.conf
index 80f5028f..0f91fae2 100644
--- a/data/gnome.session.conf
+++ b/data/gnome.session.conf
@@ -1,3 +1,48 @@
 [Unit]
-Requires=gnome-session-services.target
-Requires=org.gnome.Shell@user.service
\ No newline at end of file
+# Basic services required by any GNOME session
+Requires=gnome-session-basic-services.target
+
+# Manages monitor color calibration and night-light
+Wants=org.gnome.SettingsDaemon.Color.target
+
+# Manages location-based timezone
+Wants=org.gnome.SettingsDaemon.Datetime.target
+
+# Notifies when disk is full, notifies when RAM is full,
+# cleans up trash and temp files, donation reminders
+Wants=org.gnome.SettingsDaemon.Housekeeping.target
+
+# Notifies when events happen with printers: low on ink,
+# jammed, job cancelled, etc etc etc
+Wants=org.gnome.SettingsDaemon.PrintNotifications.target
+
+# org.freedesktop.ScreenSaver API
+Wants=org.gnome.SettingsDaemon.ScreensaverProxy.target
+
+# Flushes PulseAudio sample cache whenever the
+# sound theme changes
+Wants=org.gnome.SettingsDaemon.Sound.target
+
+# Manages SIM cards (unlocking, etc)
+Wants=org.gnome.SettingsDaemon.Wwan.target
+
+# Monitors disks for SMART errors, and notifies
+# the user if their disk is about to fail
+Wants=gnome-disk-utility-notify.service
+
+# Filesystem indexer
+Wants=localsearch-3.service
+
+# Graphical wrapper around ssh-agent
+Wants=gcr-ssh-agent.socket
+
+# Checking for automatic updates, etc
+Wants=gnome-software.service
+
+# Translates the user's standard directories
+# (Downloads/, Documents/, Music/, etc) whenever
+# the user's locale changes
+Wants=user-dirs-update-gtk.service
+
+# Notifies the user about their calendar events
+Wants=evolution-alarm-notify.service
diff --git a/data/meson.build b/data/meson.build
index 3e5401a0..76b15e92 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -1,19 +1,64 @@
-# Install the .desktop file that tells GDM about the GNOME session
+# Install the .desktop files that tell GDM about the GNOME session
 
-i18n.merge_file(
-  type: 'desktop',
-  input: configure_file(
-    input: 'gnome.desktop.in.in',
-    output: '@BASENAME@',
+wayland_desktops = [
+  'gnome',
+  'gnome-wayland',
+]
+foreach name : wayland_desktops
+  desktop = name + '.desktop'
+  desktop_in = configure_file(
+    input: desktop + '.in.in',
+    output: desktop + '.in',
     configuration: {
       'bindir': session_bindir,
+      'canrunheadless': 'true',
     },
-  ),
-  output: '@BASENAME@',
-  po_dir: po_dir,
-  install: true,
-  install_dir: session_datadir / 'wayland-sessions',
-)
+  )
+  desktop_target = i18n.merge_file(
+    type: 'desktop',
+    input: desktop_in,
+    output: desktop,
+    po_dir: po_dir,
+    install: true,
+    install_dir: session_datadir / 'wayland-sessions',
+  )
+endforeach
+
+if have_x11
+  xorg_desktops = [
+    'gnome',
+    'gnome-xorg',
+  ]
+  foreach name : xorg_desktops
+    desktop = name + '.desktop'
+
+    # 'gnome' target name is already used for wayland desktop,
+    # use 'gnome-copy' and on meson_post_install
+    # modify its name to 'gnome' again
+    if name == 'gnome'
+      desktop_out = 'gnome-copy.desktop'
+    else
+      desktop_out = desktop
+    endif
+
+    desktop_in = configure_file(
+      input: desktop + '.in.in',
+      output: desktop_out + '.in',
+      configuration: {
+        'bindir': session_bindir,
+        'canrunheadless': 'false',
+      },
+    )
+    desktop_target = i18n.merge_file(
+      type: 'desktop',
+      input: desktop_in,
+      output: desktop_out,
+      po_dir: po_dir,
+      install: true,
+      install_dir: session_datadir / 'xsessions',
+    )
+  endforeach
+endif
 
 # Next, let's install the necessary systemd scaffolding
 
@@ -42,10 +87,11 @@ foreach service: systemd_service
 endforeach
 
 systemd_target = [
+  'gnome-session-wayland@.target',
+  'gnome-session-wayland.target',
   'gnome-session@.target',
   'gnome-session.target',
   'gnome-session-basic-services.target',
-  'gnome-session-services.target',
   'gnome-session-pre.target',
   'gnome-session-manager.target',
   'gnome-session-initialized.target',
@@ -53,6 +99,12 @@ systemd_target = [
   'gnome-session-x11-services.target',
   'gnome-session-x11-services-ready.target',
 ]
+if have_x11
+  systemd_target += [
+    'gnome-session-x11@.target',
+    'gnome-session-x11.target',
+  ]
+endif
 
 install_data(
   systemd_target,
diff --git a/doc/man/gnome-session.xml b/doc/man/gnome-session.xml
index a0cff8f7..542c7938 100644
--- a/doc/man/gnome-session.xml
+++ b/doc/man/gnome-session.xml
@@ -44,6 +44,8 @@
       (either GDM, SDDM, lightdm, or other). It will load either your saved
       session, or it will provide a default session for the user as defined by
       the system administrator (or the default GNOME installation on your system).
+      Note that <command>gnome-session</command> is a wrapper script for
+      <command>gnome-session-binary</command>, which does the bulk of the work.
     </para>
     <para>
       The default session is defined in <filename>gnome.session</filename>, a
@@ -58,6 +60,15 @@
     <title>Options</title>
     <variablelist>
 
+    <varlistentry>
+      <term><option>--autostart</option>=DIR</term>
+      <listitem><para>
+        The directory <filename>DIR</filename> to be searched for autostart
+        .desktop files. This option can be used multiple times. When this option
+        is present, then default autostart directories will not be searched.
+      </para></listitem>
+    </varlistentry>
+
     <varlistentry>
       <term><option>--session</option>=SESSION</term>
       <listitem><para>
@@ -66,19 +77,87 @@
       </para></listitem>
     </varlistentry>
 
+    <varlistentry>
+      <term><option>--failsafe</option></term>
+      <listitem><para>
+        Run in fail-safe mode. User-specified applications will not be started.
+      </para></listitem>
+    </varlistentry>
+
     <varlistentry>
       <term><option>--debug</option></term>
       <listitem><para>Enable debugging code.</para></listitem>
     </varlistentry>
+
+    <varlistentry>
+      <term><option>--whale</option></term>
+      <listitem><para>
+        Show the fail whale in a dialog for debugging it.
+      </para></listitem>
+      </varlistentry>
     </variablelist>
   </refsect1>
 
+
+  <refsect1>
+    <title>Session definition</title>
+    <para>
+      Sessions are defined in <filename>.session</filename> files, that are
+      using a .desktop-like format, with the following keys in the
+      <literal>GNOME Session</literal> group:
+    </para>
+
+    <variablelist>
+      <varlistentry>
+        <term><varname>Name</varname></term>
+        <listitem><para>
+          Name of the session. This can be localized.
+        </para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>RequiredComponents</varname></term>
+        <listitem><para>
+          List of component identifiers (desktop files) that are required by
+          the session. The required components will always run in the session.
+        </para></listitem>
+      </varlistentry>
+    </variablelist>
+
+    <para>Here is an example of a session definition:</para>
+
+    <programlisting>[GNOME Session]
+Name=GNOME
+RequiredComponents=gnome-shell;gnome-settings-daemon;
+</programlisting>
+
+    <para>
+      In <command>systemd</command> managed sessions, the RequiredComponents may
+      be provided by systemd units instead. See the systemd section for more
+      information on how this works.
+    </para>
+
+    <para>
+      The <filename>.session</filename> files are looked for in
+      <filename>$XDG_CONFIG_HOME/gnome-session/sessions</filename>,
+      <filename>$XDG_CONFIG_DIRS/gnome-session/sessions</filename> and
+      <filename>$XDG_DATA_DIRS/gnome-session/sessions</filename>.
+    </para>
+  </refsect1>
+
   <refsect1><title>systemd</title>
     <para>
-      <command>gnome-session</command> uses <command>systemd</command> for the
-      bulk of its service management. <command>systemd</command> provides the
-      two special targets for desktop integration: <filename>graphical-session.target</filename>
-      and <filename>graphical-session-pre.target</filename>. These work across desktop
+      <command>gnome-session</command> will much of the session management over
+      to systemd, if systemd is available. In this case, startup components that
+      have <option>X-GNOME-HiddenUnderSystemd=true</option> set in their
+      <filename>.desktop</filename> file will be ignored by <command>gnome-session</command>.
+      It instead relies on the fact that these components are managed by systemd.
+    </para>
+
+    <para>
+      <command>systemd</command> provides the two special targets for desktop
+      integration: <filename>graphical-session.target</filename> and
+      <filename>graphical-session-pre.target</filename>. These work across desktop
       environments, and should be used if possible. <command>gnome-session</command>
       provides the following additional GNOME-specific targets:
     </para>
@@ -101,19 +180,31 @@
       </varlistentry>
 
       <varlistentry>
-        <term><filename>gnome-session@SESSION.target</filename></term>
+        <term>
+          <filename>gnome-session-x11@SESSION.target</filename>,
+          <filename>gnome-session-wayland@SESSION.target</filename>
+        </term>
+        <listitem><para>
+          Main unit started for X11/wayland based session.
+          <varname>SESSION</varname> is set according to the session that is
+          passed in via <option>--session</option>.
+        </para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term>
+          <filename>gnome-session-x11.target</filename>,
+          <filename>gnome-session-wayland.target</filename>
+        </term>
         <listitem><para>
-          Main unit started for the session. <varname>SESSION</varname> is set
-          according to the session that is passed in via <option>--session</option>.
+            Convenience units without the session embedded into the target.
         </para></listitem>
       </varlistentry>
 
       <varlistentry>
-        <term><filename>gnome-session-basic-services.target</filename></term>
+        <term><filename>gnome-session@SESSION.target</filename></term>
         <listitem><para>
-          Special unit that lists basic services that are needed by any GNOME
-          session built around <command>gnome-shell</command>. This allows for
-          convenient reuse by kiosk sessions like the GDM login screen.
+          Convenience unit with just the <varname>SESSION</varname> information embedded.
         </para></listitem>
       </varlistentry>
 
@@ -125,6 +216,7 @@
           environment variable instead of <varname>DISPLAY</varname>.
         </para></listitem>
       </varlistentry>
+
     </variablelist>
 
     <para>
@@ -142,63 +234,42 @@
   </refsect1>
 
   <refsect1>
-    <title>Session definition</title>
+    <title>Environment</title>
     <para>
-      Sessions are defined in <filename>.session</filename> files, that are
-      using a .desktop-like format, with the following keys in the
-      <literal>GNOME Session</literal> group:
+      <command>gnome-session</command> sets several environment variables for
+      use by the session:
     </para>
 
     <variablelist>
       <varlistentry>
-        <term><varname>Name</varname></term>
+        <term><filename>DISPLAY</filename></term>
         <listitem><para>
-          Name of the session. This can be localized.
+          This variable is set to the X display being used by
+          <command>gnome-session</command>. Note that if the <option>--display</option>
+          option is used this might be different from the setting of the environment
+          variable in <command>gnome-session</command>'s environment.
         </para></listitem>
       </varlistentry>
+    </variablelist>
+
+    <para>
+      Behavior of <command>gnome-session</command> ifself can be modified via
+      the following environment variables:
+    </para>
 
+    <variablelist>
       <varlistentry>
-        <term><varname>Kiosk</varname></term>
+        <term><varname>GNOME_SESSION_AUTOSTART_DIR</varname></term>
         <listitem><para>
-          Kiosk sessions are special single-purpose sessions that are built around
-          starting some specific app. For instance, a kiosk session may run a
-          full-screen web browser that's locked to a specific page. In GNOME,
-          we use kiosk sessions for the login screen (where the app is the login
-          screen UI) and the initial-setup first boot experience. When set to true,
-          <command>gnome-session</command> will make no attempt to autostart
-          any <filename>.desktop</filename> files, and will instead require that
-          the session configures all required functionality via systemd.
+          This variable specifies a list of directories to the searched for
+          autostart files. This variable overrides all directories specified
+          via the <option>--autostart</option> option, as well as all default autostart directories.
         </para></listitem>
       </varlistentry>
     </variablelist>
-
-    <para>Here is an example of a session definition:</para>
-
-    <programlisting>[GNOME Session]
-Name=GNOME
-</programlisting>
-
-    <para>
-      The <filename>.session</filename> files are loaded from
-      <filename>$XDG_CONFIG_HOME/gnome-session/sessions</filename>,
-      <filename>$XDG_CONFIG_DIRS/gnome-session/sessions</filename> and
-      <filename>$XDG_DATA_DIRS/gnome-session/sessions</filename>.
-    </para>
-
-    <para>
-      Session definitions don't do anything on their own, and they need to be
-      accompanied by <command>systemd</command> configuration to provide some
-      functionality (i.e. launch services). The recommended way to do that is
-      by hooking into one of the targets described above. For instance: when
-      defining a session named <literal>foobar</literal>, you can create
-      a systemd drop-in file in
-      <filename>/etc/systemd/user/gnome-session@foobar.target.d/session.conf</filename>
-      and define <literal>Requires=</literal> and <literal>Wants=</literal>
-      dependencies to tell systemd what services it needs to start when launching
-      this session.
-    </para>
   </refsect1>
 
+
   <refsect1>
     <title>Files</title>
 
@@ -209,10 +280,7 @@ Name=GNOME
     </literallayout></para>
     <para>
       Applications defined via <filename>.desktop</filename> files in those
-      directories will be started on login. Note that previous versions of
-      <command>gnome-session</command> would launch services from here, but
-      this is no longer the case. Services should now be configured via
-      <command>systemd</command> unit files
+      directories will be started on login.
     </para>
 
     <para><literallayout>
@@ -224,6 +292,13 @@ Name=GNOME
       These directories contain the <filename>.session</filename> files that
       can be used with the <option>--session</option> option.
     </para>
+
+    <para><literallayout>
+      <filename>$XDG_CONFIG_HOME/gnome-session/saved-sessions</filename>
+    </literallayout></para>
+    <para>
+      This directory contains the list of applications of the saved session.
+    </para>
   </refsect1>
 
   <refsect1>
diff --git a/gnome-session/gsm-app.c b/gnome-session/gsm-app.c
index b852cb26..d38138be 100644
--- a/gnome-session/gsm-app.c
+++ b/gnome-session/gsm-app.c
@@ -27,6 +27,9 @@
 #include <gio/gio.h>
 #include <gio/gdesktopappinfo.h>
 
+#define GNOME_DESKTOP_USE_UNSTABLE_API
+#include <libgnome-desktop/gnome-systemd.h>
+
 #include "gsm-app.h"
 #include "gsm-util.h"
 
@@ -189,14 +192,73 @@ gsm_app_peek_is_disabled (GsmApp *app)
         return FALSE;
 }
 
+static void
+app_launched (GAppLaunchContext *ctx,
+              GAppInfo          *appinfo,
+              GVariant          *platform_data,
+              gpointer           data)
+{
+        GsmApp *app = data;
+
+        gint pid = 0;
+        g_variant_lookup (platform_data, "pid", "i", &pid);
+
+        /* If pid == 0 the application was launched through D-Bus
+         * activation, therefore it's already in its own unit */
+        if (pid == 0)
+                return;
+
+        /* We are not interested in the result. */
+        gnome_start_systemd_scope (gsm_app_peek_app_id (app),
+                                   pid,
+                                   NULL,
+                                   NULL,
+                                   NULL, NULL, NULL);
+}
+
 gboolean
 gsm_app_start (GsmApp  *app,
                GError **error)
 {
+        g_autoptr (GAppLaunchContext) ctx = NULL;
+        GError *local_error = NULL;
+        const char * const *variable_denylist;
+        const char * const *child_environment;
+        guint handler;
+        gboolean success;
+
         g_debug ("GsmApp: starting %s (command=%s)",
                  gsm_app_peek_app_id (app),
                  g_app_info_get_commandline (G_APP_INFO (app->inner)));
-        return gsm_util_launch_app (app->inner, error);
+
+        ctx = g_app_launch_context_new ();
+
+        variable_denylist = gsm_util_get_variable_blacklist ();
+        for (size_t i = 0; variable_denylist[i] != NULL; i++)
+                g_app_launch_context_unsetenv (ctx, variable_denylist[i]);
+
+        child_environment = gsm_util_listenv ();
+        for (size_t i = 0; child_environment[i] != NULL; i++) {
+                g_auto (GStrv) split = g_strsplit (child_environment[i], "=", 2);
+                if (split[1] != NULL)
+                        g_app_launch_context_setenv (ctx, split[0], split[1]);
+        }
+
+        handler = g_signal_connect (ctx, "launched", G_CALLBACK (app_launched), app);
+        success = g_desktop_app_info_launch_uris_as_manager (app->inner,
+                                                             NULL,
+                                                             ctx,
+                                                             G_SPAWN_DO_NOT_REAP_CHILD | G_SPAWN_SEARCH_PATH,
+                                                             NULL, NULL, NULL, NULL,
+                                                             &local_error);
+        if (!success)
+                g_propagate_prefixed_error (error, local_error,
+                                            "Unable to start app (%s): ",
+                                            gsm_app_peek_app_id (app));
+
+        g_signal_handler_disconnect (ctx, handler);
+
+        return success;
 }
 
 GsmApp *
diff --git a/gnome-session/gsm-manager.c b/gnome-session/gsm-manager.c
index fecde253..b7f1788e 100644
--- a/gnome-session/gsm-manager.c
+++ b/gnome-session/gsm-manager.c
@@ -48,7 +48,6 @@
 #include "gsm-client.h"
 #include "gsm-inhibitor.h"
 #include "gsm-presence.h"
-#include "gsm-session-save.h"
 #include "gsm-shell.h"
 #include "gsm-store.h"
 #include "gsm-system.h"
@@ -60,7 +59,6 @@
 
 #define SESSION_SCHEMA            "org.gnome.desktop.session"
 #define KEY_IDLE_DELAY            "idle-delay"
-#define KEY_SAVE_RESTORE          "save-restore"
 
 #define GSM_MANAGER_SCHEMA        "org.gnome.SessionManager"
 #define KEY_LOGOUT_PROMPT         "logout-prompt"
@@ -87,7 +85,6 @@ struct _GsmManager
         GsmInhibitorFlag        inhibited_actions;
         GsmStore               *apps;
         GsmPresence            *presence;
-        GsmSessionSave         *session_save;
         char                   *session_name;
 
         /* Current status */
@@ -318,12 +315,9 @@ _start_app (const char *id,
 static void
 do_phase_startup (GsmManager *manager)
 {
-        if (manager->session_save && gsm_session_save_restore (manager->session_save))
-                g_info ("Restored from save, skipping normal autostart apps.");
-        else
-                gsm_store_foreach (manager->apps,
-                                   (GsmStoreFunc)_start_app,
-                                   manager);
+        gsm_store_foreach (manager->apps,
+                           (GsmStoreFunc)_start_app,
+                           manager);
 
         end_phase (manager);
 }
@@ -595,9 +589,6 @@ cancel_end_session (GsmManager *manager)
                            (GsmStoreFunc)_client_cancel_end_session,
                            NULL);
 
-        if (manager->session_save)
-                gsm_session_save_unseal (manager->session_save);
-
         start_phase (manager);
 }
 
@@ -850,17 +841,12 @@ gsm_manager_start (GsmManager *manager)
 }
 
 void
-_gsm_manager_set_active_session (GsmManager *manager,
-                                 const char *session_name,
-                                 gboolean    is_kiosk)
+_gsm_manager_set_active_session (GsmManager     *manager,
+                                 const char     *session_name)
 {
         g_free (manager->session_name);
         manager->session_name = g_strdup (session_name);
 
-        if (!is_kiosk)
-                manager->session_save = gsm_session_save_new (session_name);
-
-        gsm_exported_manager_set_restore_supported (manager->skeleton, manager->session_save != NULL);
         gsm_exported_manager_set_session_name (manager->skeleton, session_name);
 }
 
@@ -1367,9 +1353,6 @@ gsm_manager_shutdown (GsmExportedManager    *skeleton,
                 return TRUE;
         }
 
-        if (manager->session_save)
-                gsm_session_save_seal (manager->session_save);
-
         task = g_task_new (manager, manager->end_session_cancellable, (GAsyncReadyCallback) complete_end_session_task, invocation);
 
         manager->pending_end_session_tasks = g_slist_prepend (manager->pending_end_session_tasks,
@@ -1408,9 +1391,6 @@ gsm_manager_reboot (GsmExportedManager    *skeleton,
                 return TRUE;
         }
 
-        if (manager->session_save)
-                gsm_session_save_seal (manager->session_save);
-
         task = g_task_new (manager, manager->end_session_cancellable, (GAsyncReadyCallback) complete_end_session_task, invocation);
 
         manager->pending_end_session_tasks = g_slist_prepend (manager->pending_end_session_tasks,
@@ -1650,9 +1630,6 @@ gsm_manager_logout (GsmManager *manager,
                 return FALSE;
         }
 
-        if (manager->session_save && logout_mode != GSM_MANAGER_LOGOUT_MODE_FORCE)
-                gsm_session_save_seal (manager->session_save);
-
         manager->logout_mode = logout_mode;
 
         if (manager->phase >= GSM_MANAGER_PHASE_QUERY_END_SESSION) {
@@ -1747,92 +1724,6 @@ gsm_manager_unregister_client (GsmExportedManager    *skeleton,
         return TRUE;
 }
 
-static gboolean
-gsm_manager_register_restore (GsmExportedManager    *skeleton,
-                              GDBusMethodInvocation *invocation,
-                              const char            *app_id,
-                              const char            *dbus_name,
-                              GsmManager            *manager)
-{
-        GsmRestoreReason reason;
-        char *instance_id = NULL;
-        g_auto (GStrv) cleanup_ids = NULL;
-        gboolean valid = TRUE;
-
-        if (IS_STRING_EMPTY (dbus_name))
-                dbus_name = g_dbus_method_invocation_get_sender (invocation);
-
-        if (manager->session_save) {
-                g_debug ("GsmManager: RegisterRestore %s", app_id);
-                valid = gsm_session_save_register (manager->session_save, app_id,
-                                                   dbus_name, &reason, &instance_id,
-                                                   &cleanup_ids);
-        } else {
-                g_debug ("GsmManager: RegisterRestore %s (ignoring due to kiosk session)",
-                         app_id);
-                reason = GSM_RESTORE_REASON_PRISTINE;
-                instance_id = "";
-        }
-
-        if (valid) {
-                const char *const *cleanup = (const char**) cleanup_ids ?:
-                                             (const char *[]) { NULL };
-                gsm_exported_manager_complete_register_restore (skeleton, invocation,
-                                                                reason, instance_id,
-                                                                cleanup);
-        } else
-                g_dbus_method_invocation_return_error (invocation,
-                                                       G_DBUS_ERROR,
-                                                       G_DBUS_ERROR_INVALID_ARGS,
-                                                       "Invalid app id specified");
-        return TRUE;
-}
-
-static gboolean
-gsm_manager_deleted_instance_ids (GsmExportedManager    *skeleton,
-                                  GDBusMethodInvocation *invocation,
-                                  const char            *app_id,
-                                  const char *const     *ids,
-                                  GsmManager            *manager)
-{
-        if (manager->session_save) {
-                g_debug ("GsmManager: DeletedInstanceIds (%s, ...)", app_id);
-                gsm_session_save_deleted_ids (manager->session_save, app_id, ids);
-        } else {
-                g_debug ("GsmManager: DeletedInstanceIds (ignoring due to kiosk session)");
-        }
-
-        gsm_exported_manager_complete_deleted_instance_ids (skeleton, invocation);
-        return TRUE;
-}
-
-static gboolean
-gsm_manager_unregister_restore (GsmExportedManager    *skeleton,
-                                GDBusMethodInvocation *invocation,
-                                const char            *app_id,
-                                const char            *instance_id,
-                                GsmManager            *manager)
-{
-        gboolean found;
-
-        if (manager->session_save) {
-                g_debug ("GsmManager: UnregisterRestore (%s, instance %s)", app_id, instance_id);
-                found = gsm_session_save_unregister (manager->session_save, app_id, instance_id);
-        } else {
-                g_debug ("GsmManager: UnregisterRestore (ignoring due to kiosk session)");
-                found = TRUE; /* Not an error for an app to try and use save/restore in a kiosk session */
-        }
-
-        if (found)
-                gsm_exported_manager_complete_unregister_restore (skeleton, invocation);
-        else
-                g_dbus_method_invocation_return_error (invocation,
-                                                       GSM_MANAGER_ERROR,
-                                                       GSM_MANAGER_ERROR_NOT_REGISTERED,
-                                                       "Provided instance not found");
-        return TRUE;
-}
-
 static gboolean
 gsm_manager_inhibit (GsmExportedManager    *skeleton,
                      GDBusMethodInvocation *invocation,
@@ -2050,10 +1941,6 @@ register_manager (GsmManager *manager)
                           G_CALLBACK (gsm_manager_reboot), manager);
         g_signal_connect (skeleton, "handle-register-client",
                           G_CALLBACK (gsm_manager_register_client), manager);
-        g_signal_connect (skeleton, "handle-register-restore",
-                          G_CALLBACK (gsm_manager_register_restore), manager);
-        g_signal_connect (skeleton, "handle-deleted-instance-ids",
-                          G_CALLBACK (gsm_manager_deleted_instance_ids), manager);
         g_signal_connect (skeleton, "handle-set-reboot-to-firmware-setup",
                           G_CALLBACK (gsm_manager_set_reboot_to_firmware_setup), manager);
         g_signal_connect (skeleton, "handle-setenv",
@@ -2066,8 +1953,6 @@ register_manager (GsmManager *manager)
                           G_CALLBACK (gsm_manager_uninhibit), manager);
         g_signal_connect (skeleton, "handle-unregister-client",
                           G_CALLBACK (gsm_manager_unregister_client), manager);
-        g_signal_connect (skeleton, "handle-unregister-restore",
-                          G_CALLBACK (gsm_manager_unregister_restore), manager);
 
         manager->dbus_disconnected = FALSE;
         g_signal_connect (connection, "closed",
@@ -2201,28 +2086,19 @@ on_shell_end_session_dialog_canceled (GsmShell   *shell,
 
 static void
 handle_end_session_dialog_confirmation (GsmManager           *manager,
-                                        GsmManagerLogoutType  confirmed_logout_type)
+                                        GsmManagerLogoutType  logout_type)
 {
         disconnect_shell_dialog_signals (manager);
 
         if (manager->phase >= GSM_MANAGER_PHASE_END_SESSION)
                 return; /* Already logging out, nothing to do */
 
-        if (manager->logout_type == GSM_MANAGER_LOGOUT_SHUTDOWN &&
-            confirmed_logout_type == GSM_MANAGER_LOGOUT_REBOOT)
-                /* When triggering an offline update, gnome-shell will promote a
-                 * shutdown to a reboot. The offline update occurs after the reboot,
-                 * and then once that's done the shutdown will occur. */
-                g_info ("GsmManager: Shell converted shutdown into reboot (offline update?)");
-        else if (manager->logout_type != confirmed_logout_type)
+        if (manager->logout_type != logout_type) {
                 g_warning ("GsmManager: Shell confirmed unexpected logout type");
-
-        if (manager->session_save &&
-            !g_settings_get_boolean (manager->session_settings, KEY_SAVE_RESTORE))
-                gsm_session_save_discard (manager->session_save);
+                return; /* Make it obvious that something went wrong */
+        }
 
         manager->logout_mode = GSM_MANAGER_LOGOUT_MODE_FORCE;
-        manager->logout_type = confirmed_logout_type;
         end_phase (manager);
 }
 
diff --git a/gnome-session/gsm-manager.h b/gnome-session/gsm-manager.h
index 97b698ac..b358d2b5 100644
--- a/gnome-session/gsm-manager.h
+++ b/gnome-session/gsm-manager.h
@@ -80,8 +80,7 @@ gboolean            gsm_manager_add_legacy_session_apps        (GsmManager     *
 void                gsm_manager_start                          (GsmManager     *manager);
 
 void                _gsm_manager_set_active_session            (GsmManager     *manager,
-                                                                const char     *session_name,
-                                                                gboolean        is_kiosk);
+                                                                const char     *session_name);
 
 gboolean            gsm_manager_logout                         (GsmManager     *manager,
                                                                 guint           logout_mode,
diff --git a/gnome-session/gsm-session-fill.c b/gnome-session/gsm-session-fill.c
index a3fdc58f..b94dec25 100644
--- a/gnome-session/gsm-session-fill.c
+++ b/gnome-session/gsm-session-fill.c
@@ -34,6 +34,9 @@ load_standard_apps (GsmManager *manager,
         const char * const *system_dirs;
         int    i;
 
+        if (g_key_file_get_boolean (keyfile, GSM_KEYFILE_SESSION_GROUP, "Kiosk", NULL))
+                return;
+
         dir = g_build_filename (g_get_user_config_dir (), "autostart", NULL);
         gsm_manager_add_autostart_apps_from_dir (manager, dir);
         g_free (dir);
@@ -140,20 +143,18 @@ gboolean
 gsm_session_fill (GsmManager  *manager,
                   const char  *session)
 {
-        g_autoptr (GKeyFile) keyfile = NULL;
-        gboolean is_kiosk;
+        GKeyFile *keyfile;
 
         keyfile = get_session_keyfile (session);
+
         if (!keyfile)
                 return FALSE;
 
-        is_kiosk = g_key_file_get_boolean (keyfile, GSM_KEYFILE_SESSION_GROUP,
-                                           "Kiosk", NULL);
+        _gsm_manager_set_active_session (manager, session);
 
-        _gsm_manager_set_active_session (manager, session, is_kiosk);
+        load_standard_apps (manager, keyfile);
 
-        if (!is_kiosk)
-                load_standard_apps (manager, keyfile);
+        g_key_file_free (keyfile);
 
         return TRUE;
 }
diff --git a/gnome-session/gsm-session-save.c b/gnome-session/gsm-session-save.c
deleted file mode 100644
index 49ee1e34..00000000
--- a/gnome-session/gsm-session-save.c
+++ /dev/null
@@ -1,754 +0,0 @@
-/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
- * SPDX-License-Identifier: GPL-2.0-or-later
- */
-
-#ifdef HAVE_CONFIG_H
-#include "config.h"
-#endif
-
-#include <glib.h>
-#include <gio/gio.h>
-#include <gio/gdesktopappinfo.h>
-
-#include "gsm-session-save.h"
-#include "gsm-util.h"
-
-/* We're not generating quite so many instance IDs that we'll actually need the
- * entire length of the SHA 256 hash. So let's follow in git's footsteps and
- * truncate the instance ID to a more manageable length. It's still vanishingly
- * unlikely to run into a collision, given that apps won't have more than a couple
- * dozen instances running (and that's on the extreme high end...) */
-#define INSTANCE_ID_LEN 10
-
-struct _GsmSessionSave
-{
-        GObject          parent;
-
-        char *session_id;
-        char *file_path;
-        gboolean sealed;
-        GHashTable *apps;
-};
-
-enum {
-        PROP_0,
-        PROP_SESSION_ID,
-        LAST_PROP
-};
-
-G_DEFINE_TYPE (GsmSessionSave, gsm_session_save, G_TYPE_OBJECT);
-
-typedef struct {
-        char *app_id;
-        GDesktopAppInfo *info;
-
-        GSList *instances;
-
-        GSList *discarded_ids;
-        char *next_id;
-} SavedApp;
-
-typedef enum {
-        CRASH_STATE_NO_CRASH = 0,
-        CRASH_STATE_INSTANCE_CRASHED,
-        CRASH_STATE_SESSION_CRASHED,
-} CrashState;
-
-typedef struct {
-        char *instance_id;
-        char *dbus_name;
-        guint watch_id;
-        CrashState crash_state;
-} SavedAppInstance;
-
-static SavedApp *
-saved_app_new (const char *app_id)
-{
-        g_autofree char *desktop_id = NULL;
-        GDesktopAppInfo *info;
-        SavedApp *app;
-
-        desktop_id = g_strconcat (app_id, ".desktop", NULL);
-        info = g_desktop_app_info_new (desktop_id);
-        if (!info)
-                return NULL;
-
-        app = g_new (SavedApp, 1);
-        *app = (SavedApp) {
-                .app_id = g_strdup (app_id),
-                .info = info,
-        };
-        return app;
-}
-
-static SavedAppInstance *
-saved_app_instance_new (char *instance_id)
-{
-        SavedAppInstance *instance = g_new (SavedAppInstance, 1);
-        *instance = (SavedAppInstance) {
-                .instance_id = instance_id,
-                .dbus_name = NULL,
-                .crash_state = CRASH_STATE_NO_CRASH,
-                .watch_id = 0,
-        };
-        return instance;
-}
-
-static void
-saved_app_instance_free (SavedAppInstance *instance)
-{
-        if (instance->watch_id != 0)
-                g_bus_unwatch_name (instance->watch_id);
-
-        g_free (instance->instance_id);
-        g_free (instance->dbus_name);
-        g_free (instance);
-}
-
-static void
-saved_app_discard_instance (SavedApp *app,
-                            SavedAppInstance *instance)
-{
-        /* This is a UX quirk: basically we want the next normal launch of the
-         * app to be a relaunch of the last instance the user closed. This
-         * won't be a full restore (if the app is well-behaved, anyway), and
-         * will restore only window positions of the app's last instance. */
-        if (app->next_id)
-                app->discarded_ids = g_slist_prepend (app->discarded_ids,
-                                                      app->next_id);
-        app->next_id = g_steal_pointer (&instance->instance_id);
-}
-
-static void
-saved_app_free (SavedApp *app)
-{
-        g_free (app->app_id);
-        g_object_unref (app->info);
-        g_slist_free_full (app->instances,
-                           (GDestroyNotify) saved_app_instance_free);
-        g_slist_free_full (app->discarded_ids, g_free);
-        g_free (app->next_id);
-        g_free (app);
-}
-
-G_DEFINE_AUTOPTR_CLEANUP_FUNC (SavedApp, saved_app_free)
-G_DEFINE_AUTOPTR_CLEANUP_FUNC (SavedAppInstance, saved_app_instance_free)
-
-static void
-gsm_session_save_init (GsmSessionSave *save)
-{
-}
-
-static void
-gsm_session_save_set_session_id (GsmSessionSave *save, const char *id)
-{
-        g_set_str (&save->session_id, id);
-
-        g_clear_pointer (&save->file_path, g_free);
-        save->file_path = g_strconcat (g_get_user_state_dir (),
-                                       G_DIR_SEPARATOR_S,
-                                       "gnome-session@",
-                                       id,
-                                       ".state",
-                                       NULL);
-}
-
-static void
-gsm_session_save_set_property (GObject      *object,
-                               guint         prop_id,
-                               const GValue *value,
-                               GParamSpec   *pspec)
-{
-        GsmSessionSave *save = GSM_SESSION_SAVE (object);
-
-        switch (prop_id) {
-        case PROP_SESSION_ID:
-                gsm_session_save_set_session_id (save, g_value_get_string (value));
-                break;
-        default:
-                break;
-        }
-}
-
-static void
-gsm_session_save_get_property (GObject    *object,
-                               guint       prop_id,
-                               GValue     *value,
-                               GParamSpec *pspec)
-{
-        GsmSessionSave *save = GSM_SESSION_SAVE (object);
-
-        switch (prop_id) {
-        case PROP_SESSION_ID:
-                g_value_set_string (value, save->session_id);
-                break;
-        default:
-                break;
-        }
-}
-
-static void
-gsm_session_save_dispose (GObject *object)
-{
-        GsmSessionSave *save = GSM_SESSION_SAVE (object);
-
-        g_clear_pointer (&save->file_path, g_free);
-        g_clear_pointer (&save->apps, g_hash_table_unref);
-
-        G_OBJECT_CLASS (gsm_session_save_parent_class)->dispose (object);
-}
-
-static void gsm_session_save_constructed (GObject*);
-
-static void
-gsm_session_save_class_init (GsmSessionSaveClass *klass)
-{
-        GObjectClass *object_class = G_OBJECT_CLASS (klass);
-
-        object_class->set_property = gsm_session_save_set_property;
-        object_class->get_property = gsm_session_save_get_property;
-        object_class->constructed = gsm_session_save_constructed;
-        object_class->dispose = gsm_session_save_dispose;
-
-        g_object_class_install_property (object_class,
-                                         PROP_SESSION_ID,
-                                         g_param_spec_string ("session-id",
-                                                              "session-id",
-                                                              "session-id",
-                                                              "gnome",
-                                                              G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
-}
-
-static void
-flush_to_disk (GsmSessionSave *save)
-{
-        GVariantBuilder builder, instances;
-        g_autoptr (GVariant) variant = NULL;
-        g_autoptr (GError) error = NULL;
-        GHashTableIter app_iter;
-        gpointer value;
-
-        g_variant_builder_init_static (&builder, G_VARIANT_TYPE_VARDICT);
-
-        g_variant_builder_add (&builder, "{sv}", "version",
-                               g_variant_new_uint32 (1));
-        g_variant_builder_add (&builder, "{sv}", "dirty",
-                               g_variant_new_boolean (!save->sealed));
-
-        g_variant_builder_init_static (&instances, G_VARIANT_TYPE ("a{s(a{sv}aa{sv})}"));
-        g_hash_table_iter_init (&app_iter, save->apps);
-        while (g_hash_table_iter_next (&app_iter, NULL, &value)) {
-                SavedApp *app = value;
-
-                g_variant_builder_open (&instances, G_VARIANT_TYPE ("{s(a{sv}aa{sv})}"));
-                g_variant_builder_add (&instances, "s", app->app_id);
-
-                g_variant_builder_open (&instances, G_VARIANT_TYPE ("(a{sv}aa{sv})"));
-                g_variant_builder_open (&instances, G_VARIANT_TYPE ("a{sv}"));
-                if (app->next_id)
-                        g_variant_builder_add (&instances,
-                                               "{sv}", "next-instance-id",
-                                               g_variant_new_string (app->next_id));
-                if (app->discarded_ids) {
-                        GVariantBuilder discarded;
-                        g_variant_builder_init_static (&discarded, G_VARIANT_TYPE ("as"));
-                        for (const GSList *iter = app->discarded_ids; iter; iter = iter->next) {
-                                g_variant_builder_add (&discarded, "s", iter->data);
-                        }
-                        g_variant_builder_add (&instances, "{sv}", "discarded-instance-ids",
-                                               g_variant_builder_end (&discarded));
-                }
-                g_variant_builder_close (&instances);
-
-                g_variant_builder_open (&instances, G_VARIANT_TYPE ("aa{sv}"));
-                for (const GSList *iter = app->instances; iter; iter = iter->next) {
-                        SavedAppInstance *instance = iter->data;
-                        g_variant_builder_open (&instances, G_VARIANT_TYPE ("a{sv}"));
-                        g_variant_builder_add (&instances, "{sv}", "instance-id",
-                                               g_variant_new_string (instance->instance_id));
-                        g_variant_builder_add (&instances, "{sv}", "crashed",
-                                               g_variant_new_boolean (instance->crash_state == CRASH_STATE_INSTANCE_CRASHED));
-                        g_variant_builder_close (&instances);
-                }
-                g_variant_builder_close (&instances);
-
-                g_variant_builder_close (&instances);
-                g_variant_builder_close (&instances);
-        }
-        g_variant_builder_add (&builder, "{sv}", "instances",
-                               g_variant_builder_end (&instances));
-
-        variant = g_variant_ref_sink (g_variant_builder_end (&builder));
-
-        if (g_log_get_debug_enabled ())
-                g_debug ("GsmSessionSave: Flushing to disk:\n%s",
-                         g_variant_print (variant, TRUE));
-
-        if (!g_file_set_contents (save->file_path,
-                                  g_variant_get_data (variant),
-                                  g_variant_get_size (variant),
-                                  &error)) {
-                g_warning ("GsmSessionSave: Failed to write save file to disk: %s",
-                           error->message);
-                return;
-        }
-}
-
-#define malformed_save_error(error, msg) \
-        (g_set_error_literal ((error), G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Malformed save file: " msg), FALSE)
-
-static GHashTable *
-make_hash_table (void)
-{
-        return g_hash_table_new_full (g_str_hash, g_str_equal, NULL,
-                                      (GDestroyNotify) saved_app_free);
-}
-
-static gboolean
-deserialize_app (GVariantIter *iter, gboolean dirty, SavedApp **out_app)
-{
-        const char *app_id;
-        g_autoptr (GVariant) app_props_variant = NULL;
-        g_autoptr (GVariantDict) app_props = NULL;
-        g_autoptr (GVariantIter) discarded_ids = NULL;
-        g_autoptr (GVariantIter) instances = NULL;
-        GVariant *instance_variant = NULL;
-        g_autoptr (SavedApp) app = NULL;
-
-        if (!g_variant_iter_next (iter, "{&s(@a{sv}aa{sv})}", &app_id, &app_props_variant, &instances))
-                return FALSE;
-
-        app = saved_app_new (app_id);
-        if (!app) {
-                g_warning ("GsmSessionSave: It appears that saved app '%s' was uninstalled. Discarding",
-                           app_id);
-                *out_app = NULL;
-                return TRUE;
-        }
-
-        app_props = g_variant_dict_new (app_props_variant);
-        g_variant_dict_lookup (app_props, "next-instance-id", "s", &app->next_id);
-        if (g_variant_dict_lookup (app_props, "discarded-instance-ids", "as", &discarded_ids)) {
-                char *discarded_id = NULL;
-                while (g_variant_iter_loop (discarded_ids, "s", &discarded_id))
-                        app->discarded_ids = g_slist_prepend (app->discarded_ids,
-                                                              discarded_id);
-        }
-
-        while (g_variant_iter_loop (instances, "@a{sv}", &instance_variant)) {
-                g_auto (GVariantDict) dict = G_VARIANT_DICT_INIT (instance_variant);
-                g_autofree char *instance_id = NULL;
-                g_autoptr (SavedAppInstance) instance = NULL;
-                gboolean crashed;
-
-                if (!g_variant_dict_lookup (&dict, "instance-id", "s", &instance_id)) {
-                        g_warning ("GsmSessionSave: Dropping malformed instance of app %s",
-                                   app_id);
-                        continue;
-                }
-
-                instance = saved_app_instance_new (g_steal_pointer (&instance_id));
-
-                g_variant_dict_lookup (&dict, "crashed", "b", &crashed);
-                if (crashed)
-                        instance->crash_state = CRASH_STATE_INSTANCE_CRASHED;
-                else if (dirty)
-                        instance->crash_state = CRASH_STATE_SESSION_CRASHED;
-
-                app->instances = g_slist_prepend (app->instances,
-                                                  g_steal_pointer (&instance));
-        }
-
-        *out_app = g_steal_pointer (&app);
-        return TRUE;
-}
-
-static gboolean
-load_from_disk (GsmSessionSave  *save,
-                GError         **error)
-{
-        g_autoptr (GMappedFile) file = NULL;
-        g_autoptr (GVariant) variant = NULL;
-        g_autoptr (GVariantDict) dict = NULL;
-        g_autoptr (GVariant) instances = NULL;
-        guint32 version = 0;
-        gboolean dirty = TRUE;
-        g_autoptr (GVariantIter) instances_iter = NULL;
-        g_autoptr (GHashTable) apps = NULL;
-        SavedApp *deserialized_app;
-
-        file = g_mapped_file_new (save->file_path, FALSE, error);
-        if (!file) {
-                if (!g_error_matches (*error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-                        return FALSE;
-
-                g_clear_error (error);
-                return TRUE;
-        }
-
-        if (g_mapped_file_get_length (file) == 0)
-                return TRUE;
-
-        variant = g_variant_new_from_data (G_VARIANT_TYPE_VARDICT,
-                                           g_mapped_file_get_contents (file),
-                                           g_mapped_file_get_length (file),
-                                           FALSE,
-                                           NULL,
-                                           NULL);
-
-        if (!variant)
-                return malformed_save_error (error, "Failed to parse as variant");
-
-        g_variant_ref_sink (variant);
-        dict = g_variant_dict_new (variant);
-
-        if (!g_variant_dict_lookup (dict, "version", "u", &version))
-                return malformed_save_error (error, "Missing 'version' field");
-        if (version != 1) {
-                /* Not necessarily an error. This might happen if the user
-                 * downgrades the OS. Best we can do here is discard the save */
-                g_warning ("GsmSessionSave: Save file has unsupported version %u, discarding.",
-                           version);
-                return TRUE;
-        }
-
-        if (!g_variant_dict_lookup (dict, "dirty", "b", &dirty))
-                return malformed_save_error (error, "Missing 'dirty' field");
-
-        if (!g_variant_dict_lookup (dict, "instances", "@a{s(a{sv}aa{sv})}", &instances))
-                return malformed_save_error (error, "Missing 'instances' field");
-
-        apps = make_hash_table ();
-        instances_iter = g_variant_iter_new (instances);
-        while (deserialize_app (instances_iter, dirty, &deserialized_app))
-                if (deserialized_app)
-                        g_hash_table_insert (apps,
-                                             deserialized_app->app_id,
-                                             deserialized_app);
-
-        save->apps = g_steal_pointer (&apps);
-
-        if (g_log_get_debug_enabled ())
-                g_debug ("GsmSessionSave: Loaded %s saved session from disk:\n%s",
-                         dirty ? "dirty" : "clean",
-                         g_variant_print (variant, TRUE));
-
-        return TRUE;
-}
-
-static void
-gsm_session_save_constructed (GObject *object)
-{
-        GsmSessionSave *save = GSM_SESSION_SAVE (object);
-        GError *error = NULL;
-
-        if (g_mkdir_with_parents (g_get_user_state_dir (), 0755) < 0)
-                g_warning ("GsmSessionSave: Failed to ensure state dir exists: %m");
-
-        if (!load_from_disk (save, &error)) {
-                g_warning ("GsmSessionSave: Failed to load from disk, discarding: %s",
-                           error->message);
-                g_clear_error (&error);
-        }
-
-        if (!save->apps)
-                save->apps = make_hash_table ();
-}
-
-static gboolean
-launch_app_instance (SavedApp         *app,
-                     SavedAppInstance *instance)
-{
-        g_autoptr (GError) error = NULL;
-
-        if (instance->crash_state == CRASH_STATE_INSTANCE_CRASHED)
-                return FALSE;
-
-        g_debug ("GsmSessionSave: Launching instance of app (%s): %s",
-                 app->app_id,
-                 instance->instance_id);
-
-        if (!gsm_util_launch_app (app->info, &error)) {
-                g_warning ("GsmSessionSave: Failed to launch saved app (%s): %s",
-                           app->app_id,
-                           error->message);
-                return FALSE;
-        }
-
-        return TRUE;
-}
-
-gboolean
-gsm_session_save_restore (GsmSessionSave *save)
-{
-        gboolean launched_any = FALSE;
-        GHashTableIter iter;
-        gpointer value;
-
-        g_hash_table_iter_init (&iter, save->apps);
-        while (g_hash_table_iter_next (&iter, NULL, &value)) {
-                SavedApp *app = value;
-                for (const GSList *instance = app->instances; instance; instance = instance->next)
-                        launched_any |= launch_app_instance(app, instance->data);
-        }
-
-        return launched_any;
-}
-
-void
-gsm_session_save_seal (GsmSessionSave *save)
-{
-        if (save->sealed)
-                return;
-
-        save->sealed = TRUE;
-        flush_to_disk (save);
-}
-
-void
-gsm_session_save_unseal (GsmSessionSave *save)
-{
-        if (!save->sealed)
-                return;
-
-        save->sealed = FALSE;
-        flush_to_disk (save);
-}
-
-static void
-on_instance_vanished (GDBusConnection *connection,
-                      const gchar     *name,
-                      gpointer         user_data)
-{
-        SavedAppInstance *instance = user_data;
-
-        g_bus_unwatch_name (instance->watch_id);
-        instance->watch_id = 0;
-
-        instance->crash_state = CRASH_STATE_INSTANCE_CRASHED;
-        g_clear_pointer (&instance->dbus_name, g_free);
-}
-
-static char *
-next_instance_id (GsmSessionSave *save,
-                  SavedApp       *app)
-{
-        g_autoptr (GChecksum) checksum = NULL;
-        SavedAppInstance *collision = NULL;
-        guint32 random;
-
-        if (app->next_id)
-                return g_steal_pointer (&app->next_id);
-
-        checksum = g_checksum_new (G_CHECKSUM_SHA256);
-
-        while (TRUE) {
-                const char *digest;
-
-                g_checksum_update (checksum, (gpointer) "gnome-session", -1);
-                g_checksum_update (checksum, (gpointer) save->session_id, -1);
-                g_checksum_update (checksum, (gpointer) app->app_id, -1);
-                random = g_random_int ();
-                g_checksum_update (checksum, (gpointer) &random, sizeof (random));
-
-                digest = g_checksum_get_string (checksum);
-
-                /* Paranoia */
-                for (const GSList *iter = app->instances; iter; iter = iter->next) {
-                        SavedAppInstance *candidate = iter->data;
-                        if (strncmp (candidate->instance_id, digest, INSTANCE_ID_LEN) == 0) {
-                                collision = candidate;
-                                break;
-                        }
-                }
-
-                if (!collision)
-                        return g_strndup (digest, INSTANCE_ID_LEN);
-
-                g_checksum_reset (checksum);
-        }
-}
-
-gboolean
-gsm_session_save_register (GsmSessionSave    *save,
-                           const char        *app_id,
-                           const char        *dbus_name,
-                           GsmRestoreReason  *out_restore_reason,
-                           char             **out_instance_id,
-                           GStrv             *out_cleanup_ids)
-{
-        g_autoptr (GStrvBuilder) cleanup_builder = NULL;
-        SavedApp *app;
-        SavedAppInstance *instance = NULL;
-
-        app = g_hash_table_lookup (save->apps, app_id);
-        if (!app) {
-                app = saved_app_new (app_id);
-                if (!app)
-                        return FALSE;
-                g_hash_table_insert (save->apps, app->app_id, app);
-        }
-
-        for (const GSList *iter = app->instances; iter; iter = iter->next) {
-                SavedAppInstance *candidate = iter->data;
-                if (!candidate->dbus_name) { /* Check if the instance is running */
-                        instance = candidate;
-                        break;
-                }
-        }
-
-        if (!instance) {
-                char *instance_id = next_instance_id (save, app);
-                instance = saved_app_instance_new (instance_id);
-                app->instances = g_slist_prepend (app->instances, instance);
-                *out_restore_reason = GSM_RESTORE_REASON_LAUNCH;
-        } else if (instance->crash_state != CRASH_STATE_NO_CRASH) {
-                instance->crash_state = CRASH_STATE_NO_CRASH;
-                *out_restore_reason = GSM_RESTORE_REASON_RECOVER;
-        } else
-                *out_restore_reason = GSM_RESTORE_REASON_RESTORE;
-
-        instance->dbus_name = g_strdup (dbus_name);
-        instance->watch_id = g_bus_watch_name (G_BUS_TYPE_SESSION,
-                                               dbus_name,
-                                               G_BUS_NAME_WATCHER_FLAGS_NONE,
-                                               NULL,
-                                               on_instance_vanished,
-                                               instance,
-                                               NULL);
-
-        *out_instance_id = instance->instance_id;
-
-        cleanup_builder = g_strv_builder_new ();
-        for (const GSList *iter = app->discarded_ids; iter; iter = iter->next)
-                g_strv_builder_add (cleanup_builder, iter->data);
-        *out_cleanup_ids = g_strv_builder_end (cleanup_builder);
-
-        if (!save->sealed)
-                flush_to_disk (save);
-
-        return TRUE;
-}
-
-typedef gboolean (*DestroyPredicate)(void *obj, void *userdata);
-
-static GSList *
-remove_all_predicate (GSList           *head,
-                      DestroyPredicate  predicate,
-                      void             *userdata)
-{
-        GSList *current, **previous_ptr = &head;
-
-        while (*previous_ptr) {
-                current = *previous_ptr;
-
-                if (predicate (current->data, userdata)) {
-                        *previous_ptr = current->next;
-                        g_slist_free_1 (current);
-                } else {
-                        previous_ptr = &current->next;
-                }
-        }
-
-        return head;
-}
-
-static gboolean
-free_if_streq (char       *obj,
-               const char *userdata)
-{
-        if (!g_str_equal (obj, userdata))
-                return FALSE;
-
-        g_free (obj);
-        return TRUE;
-}
-
-void
-gsm_session_save_deleted_ids (GsmSessionSave    *save,
-                              const char        *app_id,
-                              const char *const *ids)
-{
-        SavedApp *app = g_hash_table_lookup (save->apps, app_id);
-        if (!app)
-                return;
-
-        for (; *ids; ids++)
-                app->discarded_ids = remove_all_predicate (app->discarded_ids,
-                                                           (DestroyPredicate) free_if_streq,
-                                                           (void*) *ids);
-
-        if (!save->sealed)
-                flush_to_disk (save);
-}
-
-gboolean
-gsm_session_save_unregister (GsmSessionSave *save,
-                             const char     *app_id,
-                             const char     *instance_id)
-{
-        SavedApp *app;
-        SavedAppInstance *instance = NULL;
-
-        app = g_hash_table_lookup (save->apps, app_id);
-        if (!app)
-                return FALSE;
-
-        for (const GSList *iter = app->instances; iter; iter = iter->next) {
-                SavedAppInstance *candidate = iter->data;
-                if (g_str_equal (instance_id, candidate->instance_id)) {
-                        instance = candidate;
-                        break;
-                }
-        }
-        if (!instance)
-                return FALSE;
-
-        saved_app_discard_instance (app, instance);
-        app->instances = g_slist_remove (app->instances, instance);
-        saved_app_instance_free (g_steal_pointer (&instance));
-
-        if (!save->sealed)
-                flush_to_disk (save);
-
-        return TRUE;
-}
-
-static gboolean
-free_if_not_crashed (SavedAppInstance *instance,
-                     SavedApp *app)
-{
-        /* Remember instances of crashed apps, so that we communicate the right
-         * restore reason on next launch */
-        if (instance->crash_state == CRASH_STATE_INSTANCE_CRASHED)
-                return FALSE;
-
-        saved_app_discard_instance (app, instance);
-        saved_app_instance_free (instance);
-        return TRUE;
-}
-
-void
-gsm_session_save_discard (GsmSessionSave *save)
-{
-        GHashTableIter iter;
-        gpointer value;
-
-        g_hash_table_iter_init (&iter, save->apps);
-        while (g_hash_table_iter_next (&iter, NULL, &value)) {
-                SavedApp *app = value;
-                app->instances = remove_all_predicate (app->instances,
-                                                       (DestroyPredicate) free_if_not_crashed,
-                                                       app);
-        }
-
-        flush_to_disk (save);
-}
-
-GsmSessionSave *
-gsm_session_save_new (const char *session_id)
-{
-        return g_object_new (GSM_TYPE_SESSION_SAVE,
-                             "session-id", session_id,
-                             NULL);
-}
diff --git a/gnome-session/gsm-session-save.h b/gnome-session/gsm-session-save.h
deleted file mode 100644
index 82ee61cb..00000000
--- a/gnome-session/gsm-session-save.h
+++ /dev/null
@@ -1,39 +0,0 @@
-/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
- * SPDX-License-Identifier: GPL-2.0-or-later
- */
-
-#pragma once
-
-#include <glib-object.h>
-
-G_BEGIN_DECLS
-
-typedef enum {
-        GSM_RESTORE_REASON_PRISTINE,
-        GSM_RESTORE_REASON_LAUNCH,
-        GSM_RESTORE_REASON_RECOVER,
-        GSM_RESTORE_REASON_RESTORE,
-} GsmRestoreReason;
-
-#define GSM_TYPE_SESSION_SAVE (gsm_session_save_get_type ())
-G_DECLARE_FINAL_TYPE (GsmSessionSave, gsm_session_save, GSM, SESSION_SAVE, GObject)
-
-GsmSessionSave *gsm_session_save_new         (const char         *session_id);
-gboolean        gsm_session_save_restore     (GsmSessionSave     *save);
-void            gsm_session_save_seal        (GsmSessionSave     *save);
-void            gsm_session_save_unseal      (GsmSessionSave     *save);
-gboolean        gsm_session_save_register    (GsmSessionSave     *save,
-                                              const char         *app_id,
-                                              const char         *dbus_name,
-                                              GsmRestoreReason   *out_restore_reason,
-                                              char              **out_instance_id,
-                                              GStrv              *out_cleanup_ids);
-void            gsm_session_save_deleted_ids (GsmSessionSave     *save,
-                                              const char         *app_id,
-                                              const char *const  *ids);
-gboolean        gsm_session_save_unregister  (GsmSessionSave     *save,
-                                              const char         *app_id,
-                                              const char         *instance_id);
-void     gsm_session_save_discard            (GsmSessionSave *save);
-
-G_END_DECLS
diff --git a/gnome-session/gsm-util.c b/gnome-session/gsm-util.c
index f2cb6168..9dd3e8e1 100644
--- a/gnome-session/gsm-util.c
+++ b/gnome-session/gsm-util.c
@@ -30,9 +30,6 @@
 #include <glib/gstdio.h>
 #include <gio/gio.h>
 
-#define GNOME_DESKTOP_USE_UNSTABLE_API
-#include <libgnome-desktop/gnome-systemd.h>
-
 #include "gsm-util.h"
 
 static gchar **child_environment;
@@ -486,64 +483,3 @@ gsm_util_get_variable_blacklist (void)
 {
         return variable_blacklist;
 }
-
-static void
-app_launched (GAppLaunchContext *ctx,
-              GAppInfo          *appinfo,
-              GVariant          *platform_data,
-              gpointer           data)
-{
-        gint pid = 0;
-        g_variant_lookup (platform_data, "pid", "i", &pid);
-
-        /* If pid == 0 the application was launched through D-Bus
-         * activation, therefore it's already in its own unit */
-        if (pid == 0)
-                return;
-
-        /* We are not interested in the result. */
-        gnome_start_systemd_scope (g_app_info_get_id (appinfo),
-                                   pid, NULL, NULL, NULL, NULL, NULL);
-}
-
-gboolean
-gsm_util_launch_app (GDesktopAppInfo *app,
-                     GError         **error)
-{
-        g_autoptr (GAppLaunchContext) ctx = NULL;
-        GError *local_error = NULL;
-        const char * const *variable_denylist;
-        const char * const *child_environment;
-        guint handler;
-        gboolean success;
-
-        ctx = g_app_launch_context_new ();
-
-        variable_denylist = gsm_util_get_variable_blacklist ();
-        for (size_t i = 0; variable_denylist[i] != NULL; i++)
-                g_app_launch_context_unsetenv (ctx, variable_denylist[i]);
-
-        child_environment = gsm_util_listenv ();
-        for (size_t i = 0; child_environment[i] != NULL; i++) {
-                g_auto (GStrv) split = g_strsplit (child_environment[i], "=", 2);
-                if (split[1] != NULL)
-                        g_app_launch_context_setenv (ctx, split[0], split[1]);
-        }
-
-        handler = g_signal_connect (ctx, "launched", G_CALLBACK (app_launched), app);
-
-        success = g_desktop_app_info_launch_uris_as_manager (app,
-                                                             NULL,
-                                                             ctx,
-                                                             G_SPAWN_SEARCH_PATH,
-                                                             NULL, NULL, NULL, NULL,
-                                                             &local_error);
-        if (!success)
-                g_propagate_prefixed_error (error, local_error,
-                                            "Unable to start app (%s): ",
-                                            g_app_info_get_id (G_APP_INFO (app)));
-
-        g_signal_handler_disconnect (ctx, handler);
-
-        return success;
-}
diff --git a/gnome-session/gsm-util.h b/gnome-session/gsm-util.h
index 5c578de7..2cc129f9 100644
--- a/gnome-session/gsm-util.h
+++ b/gnome-session/gsm-util.h
@@ -19,8 +19,6 @@
 #define __GSM_UTIL_H__
 
 #include <glib.h>
-#include <gio/gio.h>
-#include <gio/gdesktopappinfo.h>
 
 G_BEGIN_DECLS
 
@@ -41,9 +39,6 @@ const char * const * gsm_util_get_variable_blacklist(void);
 gboolean    gsm_util_export_activation_environment  (GError     **error);
 gboolean    gsm_util_export_user_environment        (GError     **error);
 
-gboolean    gsm_util_launch_app                     (GDesktopAppInfo  *app,
-                                                     GError          **error);
-
 G_END_DECLS
 
 #endif /* __GSM_UTIL_H__ */
diff --git a/gnome-session/leader-main.c b/gnome-session/leader-main.c
index 8ccc06d3..091c7754 100644
--- a/gnome-session/leader-main.c
+++ b/gnome-session/leader-main.c
@@ -78,8 +78,7 @@ maybe_reexec_with_login_shell (GStrv argv)
         if (g_strcmp0 (g_getenv ("XDG_SESSION_CLASS"), "greeter") == 0)
                 return;
 
-        /* We only need to run this on Wayland. On X11 the Xsession script will
-         * do this for us, and on the TTY we'd be launched from a login shell */
+        /* Or with X (it has a mess of scripts that do this for us) */
         if (g_strcmp0 (g_getenv ("XDG_SESSION_TYPE"), "wayland") != 0)
                 return;
 
@@ -116,16 +115,10 @@ maybe_reexec_with_login_shell (GStrv argv)
 static void
 set_up_environment (void)
 {
-        g_autofree char *dbus_address = NULL;
         g_autoptr (GSettings) locale_settings = NULL;
         g_autofree char *region = NULL;
         g_autofree char *ibus_path = NULL;
 
-        dbus_address = g_dbus_address_get_for_bus_sync (G_BUS_TYPE_SESSION, NULL, NULL);
-        if (!dbus_address)
-                g_error ("No session bus running!");
-        g_setenv ("DBUS_SESSION_BUS_ADDRESS", dbus_address, TRUE);
-
         locale_settings = g_settings_new ("org.gnome.system.locale");
         region = g_settings_get_string (locale_settings, "region");
         if (!IS_STRING_EMPTY (region)) {
@@ -198,6 +191,10 @@ main (int argc, char **argv)
                 { NULL, 0, 0, 0, NULL, NULL, NULL }
         };
 
+        /* Make sure that we have a session bus */
+        if (!g_getenv ("DBUS_SESSION_BUS_ADDRESS"))
+                g_error ("No session bus running! Cannot continue");
+
         setlocale (LC_ALL, "");
         bindtextdomain (GETTEXT_PACKAGE, LOCALE_DIR);
         bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
@@ -230,11 +227,6 @@ main (int argc, char **argv)
         if (!no_reexec)
                 maybe_reexec_with_login_shell (argv_dup);
 
-        if (g_strcmp0 (g_getenv ("XDG_SESSION_TYPE"), "tty") == 0) {
-                g_warning ("Launched from the TTY. Launching a Wayland session");
-                g_setenv ("XDG_SESSION_TYPE", "wayland", TRUE);
-        }
-
         if (disable_acceleration_check)
                 g_warning ("Wayland assumes that acceleration works, so --disable-acceleration-check is deprecated!");
 
diff --git a/gnome-session/leader-systemd.c b/gnome-session/leader-systemd.c
index 911b7d84..1fc06a63 100644
--- a/gnome-session/leader-systemd.c
+++ b/gnome-session/leader-systemd.c
@@ -306,11 +306,15 @@ main (int argc, char **argv)
         if (ctx.session_bus == NULL)
                 g_error ("Failed to obtain session bus: %s", error->message);
 
-        if (systemd_unit_is_active (ctx.session_bus, "graphical-session.target", &error))
-                g_error ("A graphical session is already running!");
+        /* We don't escape the name (i.e. we leave any '-' intact). */
+        target = g_strdup_printf ("gnome-session-%s@%s.target",
+                                  g_getenv ("XDG_SESSION_TYPE"), session_name);
+
+        if (systemd_unit_is_active (ctx.session_bus, target, &error))
+                g_error ("Session manager is already running!");
         if (error != NULL)
-                g_warning ("Failed to check if a graphical session is running: %s",
-                           error->message);
+                g_warning ("Failed to check if unit %s is active: %s",
+                           target, error->message);
         g_clear_error (&error);
 
         /* Reset all failed units; we are going to start a lot of things and
@@ -320,8 +324,6 @@ main (int argc, char **argv)
                 g_warning ("Failed to reset failed state of units: %s", error->message);
         g_clear_error (&error);
 
-        /* We don't escape the name (i.e. we leave any '-' intact). */
-        target = g_strdup_printf ("gnome-session@%s.target", session_name);
         g_message ("Starting GNOME session target: %s", target);
 
         if (!systemd_start_unit (ctx.session_bus, target, "fail", &error))
diff --git a/gnome-session/meson.build b/gnome-session/meson.build
index 67774960..403d0572 100644
--- a/gnome-session/meson.build
+++ b/gnome-session/meson.build
@@ -41,7 +41,6 @@ sources = files(
   'gsm-manager.c',
   'gsm-presence.c',
   'gsm-session-fill.c',
-  'gsm-session-save.c',
   'gsm-shell.c',
   'gsm-store.c',
   'gsm-system.c',
diff --git a/gnome-session/org.gnome.SessionManager.xml b/gnome-session/org.gnome.SessionManager.xml
index 1a5bec6c..77a37ba5 100644
--- a/gnome-session/org.gnome.SessionManager.xml
+++ b/gnome-session/org.gnome.SessionManager.xml
@@ -84,103 +84,6 @@
       <arg type="o" name="client_id" direction="in"/>
     </method>
 
-    <!--
-      RegisterRestore:
-      @app_id: The application identifier
-      @dbus_name: The app's dbus unique name, for crash monitoring
-      @reason: The app's restore reason, see below
-      @instance_id: A stable ID for this instance of this app
-      @cleanup_ids: A list of instance IDs the session manager wishes to discard
-
-      Register the specified app for session save/restore functionality. Apps
-      that register here will be relaunched on next login, and can restore their
-      saved state.
-
-      The restore @reason communicates to an app how much state it should restore.
-      See the following table:
-
-      <itemizedlist>
-        <listitem><para>
-          <emphasis>0</emphasis>:
-          The app shouldn't restore any state
-        </para></listitem>
-        <listitem><para>
-          <emphasis>1</emphasis>:
-          This is a normal app launch. The app should restore as much state
-          as is reasonable. For instance: an app like GNOME Calculator may
-          restore the last-used mode (Basic/Advanced/Programmer/...) but won't
-          restore the history.
-        </para></listitem>
-        <listitem><para>
-          <emphasis>2</emphasis>:
-          This is an app restore after a crash. The app should restore as much
-          state as is possible, but should be careful about causing the crash
-          again. For instance, an app like GNOME Web might immediately load all
-          of its tabs after a normal restore, but only show the "This page may
-          have caused Web to crash" page in each tab after a crash restore. This
-          is also returned if the entire session crashed
-        </para></listitem>
-        <listitem><para>
-          <emphasis>3</emphasis>:
-          This is a normal app restore. The app was shut down due to logout or
-          some other reason, and the session was cleanly ended. The app should
-          restore as much state as is possible.
-        </para></listitem>
-      </itemizedlist>
-
-      The @instance_id incorporates a string that identifies this session. This
-      means that the app can safely assume that there will never be an instance
-      ID collision with other session managers from other desktop environments,
-      or even other sessions managed by gnome-session.
-
-      The instance IDs returned in @cleanup_ids will never be reused, and the
-      app should delete any data associated with these IDs from disk. When
-      complete, the app should notify the session manager via
-      #org.gnome.SessionManager.DeletedInstanceIds()
-
-      This is a private interface between gnome-session and xdg-desktop-portal.
-      Apps should talk to the portal directly for this functionality.
-    -->
-    <method name="RegisterRestore">
-      <arg type="s" name="app_id" direction="in"/>
-      <arg type="s" name="dbus_name" direction="in"/>
-      <arg type="u" name="reason" direction="out"/>
-      <arg type="s" name="instance_id" direction="out"/>
-      <arg type="as" name="cleanup_ids" direction="out"/>
-    </method>
-
-    <!--
-      DeletedInstanceIds:
-      @app_id: The application identifier
-      @ids: List of instance IDs the application deleted
-
-      Used to notify the session manager that the app has successfully deleted
-      the instance IDs, and that the manager is now free to forget them for good.
-
-      This is a private interface between gnome-session and xdg-desktop-portal.
-      Apps should talk to the portal directly for this functionality.
-    -->
-    <method name="DeletedInstanceIds">
-      <arg type="s" name="app_id" direction="in"/>
-      <arg type="as" name="ids" direction="in"/>
-    </method>
-
-    <!--
-      UnregisterRestore:
-      @app_id: The application identifier
-      @instance_id: The application's instance_id
-
-      Un-registers the app from the session save/restore functionality. This
-      should be called whenever the app exits cleanly.
-
-      This is a private interface between gnome-session and xdg-desktop-portal.
-      Apps should talk to the portal directly for this functionality.
-    -->
-    <method name="UnregisterRestore">
-      <arg type="s" name="app_id" direction="in"/>
-      <arg type="s" name="instance_id" direction="in"/>
-    </method>
-
     <!--
       Inhibit:
       @app_id: The application identifier, optional
@@ -453,12 +356,5 @@
     -->
     <property name="InhibitedActions" type="u" access="read"/>
 
-    <!--
-      RestoreSupported:
-
-      Communicates whether the session supports the save/restore mechanism.
-    -->
-    <property name="RestoreSupported" type="b" access="read"/>
-
   </interface>
 </node>
diff --git a/meson.build b/meson.build
index 8cef327f..b10c9455 100644
--- a/meson.build
+++ b/meson.build
@@ -1,12 +1,13 @@
 project(
   'gnome-session', 'c',
-  version: '50.beta',
+  version: '50.beta',
   license: 'GPL2+',
   default_options: 'buildtype=debugoptimized',
   meson_version: '>= 0.60.0'
 )
 
 session_version = meson.project_version()
+session_api_name = '@0@-@1@'.format(meson.project_name(), session_version.split('.')[0])
 
 session_prefix = get_option('prefix')
 session_bindir = join_paths(session_prefix, get_option('bindir'))
@@ -19,6 +20,7 @@ session_pkgdatadir = join_paths(session_datadir, meson.project_name())
 
 cc = meson.get_compiler('c')
 
+have_x11 = get_option('x11')
 config_h = configuration_data()
 
 # defines
@@ -28,7 +30,7 @@ set_defines = [
   ['PACKAGE_VERSION', session_version],
   ['VERSION', session_version],
   # i18n
-  ['GETTEXT_PACKAGE', meson.project_name()]
+  ['GETTEXT_PACKAGE', session_api_name]
 ]
 
 foreach define: set_defines
@@ -79,11 +81,13 @@ add_project_arguments(common_flags + compiler_flags, language: 'c')
 glib_req_version = '>= 2.82.0'
 
 gio_dep = dependency('gio-2.0', version: glib_req_version)
+gtk_dep = dependency('gtk4')
 
 session_deps = [
   gio_dep,
   dependency('glib-2.0', version: glib_req_version),
   dependency('gnome-desktop-4'),
+  dependency('json-glib-1.0', version: '>= 0.10')
 ]
 
 libsystemd_dep = dependency('libsystemd', version: '>= 209', required: true)
@@ -121,6 +125,13 @@ subdir('data')
 subdir('doc')
 subdir('po')
 
+if have_x11
+  meson.add_install_script(
+    'meson_post_install.py',
+    session_datadir
+  )
+endif
+
 gnome.post_install(
   glib_compile_schemas: true,
 )
diff --git a/meson_options.txt b/meson_options.txt
index e1d1afae..781d8eea 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -2,4 +2,5 @@ option('deprecation_flags', type: 'boolean', value: false, description: 'use *_D
 option('systemduserunitdir', type: 'string', description: 'Directory for systemd user service files')
 option('docbook', type: 'boolean', value: true, description: 'build documentation')
 option('man', type: 'boolean', value: true, description: 'build documentation (requires xmlto)')
+option('x11', type: 'boolean', value: false, description: 'enable x11 session support')
 option('mimeapps', type: 'boolean', value: true, description: 'install default gnome-mimeapps.list')
diff --git a/meson_post_install.py b/meson_post_install.py
new file mode 100644
index 00000000..b7945826
--- /dev/null
+++ b/meson_post_install.py
@@ -0,0 +1,14 @@
+#!/usr/bin/env python3
+
+import os
+import shutil
+import sys
+
+if os.environ.get('DESTDIR'):
+  install_root = os.environ.get('DESTDIR') + os.path.abspath(sys.argv[1])
+else:
+  install_root = sys.argv[1]
+
+src = os.path.join(install_root, 'xsessions', 'gnome-copy.desktop')
+dst = os.path.join(install_root, 'xsessions', 'gnome.desktop')
+shutil.move(src, dst)

diff --git a/tools/gnome-session-inhibit.c b/tools/gnome-session-inhibit.c
index 41218e8e..dc1ee542 100644
--- a/tools/gnome-session-inhibit.c
+++ b/tools/gnome-session-inhibit.c
@@ -382,7 +382,6 @@ int main (int argc, char *argv[])
   g_option_context_set_main_group (opts, main_opts);
   g_option_context_set_summary (opts, N_("Execute COMMAND while inhibiting some session functionality."));
   g_option_context_set_description (opts, N_("Valid ACTION values are: logout, switch-user, suspend, idle, and automount.\nIf no --inhibit options are specified, idle is assumed."));
-  g_option_context_set_strict_posix (opts, TRUE);
 
   if (!g_option_context_parse (opts, &argc, &argv, &error))
     {
