diff --git a/games/meson.build b/games/meson.build
index 992fba5..62fce6b 100644
--- a/games/meson.build
+++ b/games/meson.build
@@ -43,3 +43,13 @@ lggs = library('gnome-games-support-3',
     version: '3',
     dependencies: lggs_interface_deps,
 )
+
+pkg = import('pkgconfig')
+pkg.generate(
+    lggs,
+    filebase: 'libgnome-games-support-3',
+    requires: lggs_interface_deps,
+    name: meson.project_name(),
+    description: 'Useful functionality shared among GNOME games',
+    subdirs: includesubdir,
+)
