diff --git a/lib/serializer.vala b/lib/serializer.vala
index 8bf3573ea..0dd13c4bb 100644
--- a/lib/serializer.vala
+++ b/lib/serializer.vala
@@ -449,7 +449,24 @@ public class Serializer : Object
             }
         }
 
-        string.append (cast_to_string (mantissa, ref n_digits));
+        var mantissa_string = cast_to_string (mantissa, ref n_digits);
+        if (eng_format)
+        {
+            mantissa_string = mantissa_string.replace (tsep.to_string (), "", 1);
+            if (mantissa_string.has_prefix ("1000"))
+            {
+                exponent += 3;
+                mantissa_string = mantissa_string.splice (0, 4, "1");
+                n_digits = 1;
+            }
+        }
+        else if (mantissa_string.has_prefix ("10"))
+        {
+            exponent += 1;
+            mantissa_string = mantissa_string.splice (0, 2, "1");
+            n_digits = 1;
+        }
+        string.append (mantissa_string);
 
         return exponent;
     }
-- 
GitLab
